numpy/core/src/multiarray/arraytypes.c.src

```diff
@@ -3692,9 +3691,12 @@ static void
         }
     }
     else {
-        for (i = 0; i < ni; i++) {
+        for (i = 0, j = 0; i < ni; i++, j++) {
+            if (j >= nv) {
+                j = 0;
+            }
             if (mask[i]) {
-                in[i] = vals[i%nv];
+                in[i] = vals[i];
```

Comment:
I did have the `nv >= ni` branch at one point, but removed it because it was showing virtually no performance improvement. Guess we now know why. ;-)